core/src/Kokkos_GraphNode.hpp

```diff
@@ -0,0 +1,469 @@
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_KOKKOS_GRAPHNODE_HPP
+#define KOKKOS_KOKKOS_GRAPHNODE_HPP
+
+#include <Kokkos_Macros.hpp>
+
+#include <impl/Kokkos_Error.hpp>  // contract macros
+
+#include <Kokkos_Core_fwd.hpp>
+#include <Kokkos_Graph_fwd.hpp>
+#include <impl/Kokkos_GraphImpl_fwd.hpp>
+#include <Kokkos_Parallel_Reduce.hpp>
+#include <impl/Kokkos_GraphImpl_Utilities.hpp>
+#include <impl/Kokkos_GraphImpl.hpp>  // GraphAccess
+
+#include <memory>  // std::shared_ptr
+
+namespace Kokkos {
+namespace Experimental {
+
+template <class ExecutionSpace, class Kernel /*= TypeErasedTag*/,
+          class Predecessor /*= TypeErasedTag*/>
+class GraphNodeRef {
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="template parameter constraints"> {{{2
+
+  // Note: because of these assertions, instantiating this class template is not
+  //       intended to be SFINAE-safe, so do validation before you instantiate.
+
+  static_assert(
+      std::is_same<Predecessor, TypeErasedTag>::value ||
+          Kokkos::Impl::is_specialization_of<Predecessor, GraphNodeRef>::value,
+      "Invalid predecessor template parameter given to GraphNodeRef");
+
+  static_assert(
+      Kokkos::is_execution_space<ExecutionSpace>::value,
+      "Invalid execution space template parameter given to GraphNodeRef");
+
+  static_assert(std::is_same<Predecessor, TypeErasedTag>::value ||
+                    Kokkos::Impl::is_graph_kernel<Kernel>::value,
+                "Invalid kernel template parameter given to GraphNodeRef");
+
+  static_assert(!Kokkos::Impl::is_more_type_erased<Kernel, Predecessor>::value,
+                "The kernel of a graph node can't be more type-erased than the "
+                "predecessor");
+
+  // </editor-fold> end template parameter constraints }}}2
+  //----------------------------------------------------------------------------
+
+ public:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="public member types"> {{{2
+
+  using execution_space   = ExecutionSpace;
+  using graph_kernel      = Kernel;
+  using graph_predecessor = Predecessor;
+
+  // </editor-fold> end public member types }}}2
+  //----------------------------------------------------------------------------
+
+ private:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="Friends"> {{{2
+
+  template <class, class, class>
+  friend class GraphNodeRef;
+  template <class>
+  friend struct GraphBuilder;
```

Comment:
that is gone or?

Reply:
A bunch of these can go away to be honest. Resolved in 5cb44c68